### PR TITLE
core/state: remove notion of fake storage

### PR DESCRIPTION
**Pull request by holiman**
This PR fixes https://github.com/ethereum/go-ethereum/issues/24456 . 

This PR removes the notion of `fakeStorage` from the state objects, and instead, for any state modifications that are needed, it simply makes the changes and commits it. 

I'm not fully certain if there are any tangible negative side-effects of doing a `Commit` (such as, does it have potential to flush stuff to disk? I _think_ not). 

One other change is that this change makes it not possible to replace the entire storage in one go, rather we can only update slots explicitly. I don't know how much of a biggie that is.

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -885,6 +885,8 @@ func (diff *StateOverride) Apply(state *state.StateDB) error {
 			}
 		}
 	}
+	// Now commit the changes
+	state.Commit(false)
```
Comment (karalabe) [inline internal/ethapi/api.go:889]: That will actually write into the trie database, which is a big nono. It might still be "just in ram", but it will compete with the memory cache of real trie nodes. I'd wager it also creates a diff layer on top of the snapshots if one exists.

Comment (holiman) [reply internal/ethapi/api.go]: Gotcha! Using finalize seems to work  too

File: core/state/state_object.go [hunk 2]
```diff
@@ -268,16 +254,13 @@ func (s *stateObject) SetState(db Database, key, value common.Hash) {
...
-func (s *stateObject) SetStorage(storage map[common.Hash]common.Hash) {
-	// Allocate fake storage if it's nil.
-	if s.fakeStorage == nil {
-		s.fakeStorage = make(Storage)
-	}
-	for key, value := range storage {
-		s.fakeStorage[key] = value
+//
+//@deprecated use SetState per item instead. This method no longer fully replaces
+// all slots.
```
Comment (karalabe) [inline core/state/state_object.go:259]: I'm not following this deprecation warning? It didn't fully replace all the slots before either.

Comment (karalabe) [reply core/state/state_object.go:259]: Can't we just delete this altogether if there's a better alternative?

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -885,6 +885,10 @@ func (diff *StateOverride) Apply(state *state.StateDB) error {
 			}
 		}
 	}
+	// Now finalize the changes. Finalize is normally performed between transactions.
+	// By using finalize, the overrides are semantically behaving as
+	// if they were created in a transaction just before the tracing occur.
```
Comment (karalabe) [inline internal/ethapi/api.go:890]: "before the tracing occur" -> "before the tracing occurs"

**Comment by holiman** [comment]
I've made a stab at fixing this PR now, so that it should be possible to override the whole state. I haven't added any test for it yet.

**Comment by rjl493456442** [comment]
I will take a look tomorrow.

**Comment by holiman** [comment]
Test added!

File: core/state/state_object.go [hunk 4]
```diff
@@ -269,16 +255,15 @@ func (s *stateObject) SetState(db Database, key, value common.Hash) {
...
-		s.fakeStorage = make(Storage)
-	}
-	for key, value := range storage {
-		s.fakeStorage[key] = value
+//
+func (s *stateObject) SetStorage(db Database, storage map[common.Hash]common.Hash) {
+	// We set the root to empty root, to make prevent GetCommittedState from
+	// reading the original data, it should thus only have access to what is
+	// set in the loop below.
+	s.data.Root = emptyRoot
```
Comment (rjl493456442) [inline core/state/state_object.go:263]: Hmm, it's not enough. State can still be hit in snapshot. e.g. https://github.com/ethereum/go-ethereum/blob/master/core/state/state_object.go#L216

Comment (holiman) [reply core/state/state_object.go:263]: No, I think we're executing snapshot-free when we're tracing

**Comment by holiman** [comment]
Lifting this out 

@rjl493456442 said:
> Hmm, it's not enough. State can still be hit in snapshot. e.g. https://github.com/ethereum/go-ethereum/blob/master/core/state/state_object.go#L216

@holiman said:
> No, I think we're executing snapshot-free when we're tracing

-----

The state overrides come from two places, either `ethapi/api.go` `DoCall`:
```
	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
	if state == nil || err != nil {
		return nil, err
	}
	if err := overrides.Apply(state); err != nil {

```
This one uses 
```
		stateDb, err := b.eth.BlockChain().StateAt(header.Root)
```
->
```
// StateAt returns a new mutable state based on a particular point in time.
func (bc *BlockChain) StateAt(root common.Hash) (*state.StateDB, error) {
	return state.New(root, bc.stateCache, bc.snaps)
}

```
So, if the `b.eth.Blockchain()` has snapshots, then so does the state we're tracing on. 
----------
Or `eth/tracers/api.go`, `API.TraceCall`
```
	statedb, release, err := api.backend.StateAtBlock(ctx, block, reexec, nil, true, false)
	if err != nil {
		return nil, err
	}
	defer release()

	vmctx := core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil)
	// Apply the customization rules if required.
	if config != nil {
		if err := config.StateOverrides.Apply(statedb); err != nil {
```
This one is a bit different, though, because `StateAtBlock` explicitly creates a new state without snapshots, for tracing. 

## TL;DR

For tracing, we are (probably) trie-only, and don't have to worry about snapshot. For `eth_Call` variants, however, we do, and this PR will not work properly.

**Comment by rjl493456442** [comment]
@holiman I think we shouldn't implicitly hold the assumption that snapshot is not available. Theoretically, if a state is for reading, both trie and snapshot might be available.

**Comment by holiman** [comment]
> @holiman I think we shouldn't implicitly hold the assumption that snapshot is not available. Theoretically, if a state is for reading, both trie and snapshot might be available.

Yeah, that makes a lot of sense

**Comment by holiman** [comment]
>  I think we shouldn't implicitly hold the assumption that snapshot is not available. 

I think this PR now respects this idea. One might consider the way it's done to be slightly hacky, but I really do not see a better way to achieve the same effect.

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM
